### PR TITLE
Redstone fixes + BlockEntity issues fixed

### LINKED ISSUE #727: Graphic bug by placing sticky piston on redstone block
# 🐞 I found a bug

### 📸 Screenshots / Videos

https://youtu.be/_Q37xcy-o2E

### ▶ Steps to Reproduce

1. Place a sticky piston on a redstone block.

### ✔ Expected Behavior

It remains a sticky piston

### ❌ Actual Behavior

The graphic turns to a normal piston but it has the same properties

### 📋 Debug information

* PowerNukkit version: 1.4.0.0-PN-SNAPSHOT (6458c38)
* Debug link: ✍
* Timings link (if relevant): ✍

### PR DESCRIPTION
**Pull request by PocketMiner82**
Fixes blockentity was created twice of following blocks because the `#onUpdate()` method created an entity with `#getOrCreateBlockEntity()` and the `#setBlockAndCreateEntity()`, mostly used in the `#place()` method first sets the block with update and then creates the blockentity (also fixes #727):

- [x] BlockPistonBase
- [X] BlockNoteBlock
- [X] BlockRedstoneComparator

Blockentitiy close:
- [X] If you break a piston head, the blockentity of the piston should get removed.

Fixes some redstone issues:
- [X] This should work:
Redstone_transparent_blocks
- [X] Observer should trigger these updates:
  - [X] Piston extend/retract
  - [X] String enter/leave without being attached to tripwirehook
  - [X] Tripwire hook trigger
  - [X] On dispense fail
  - [X] Comparator click or update
  - [x] Redstone lamp on/off
- [x] Observers have wrong delay; 3 redstone ticks (= 6 server ticks) delay and 1 redstone tick (= 2 server ticks) on, messed with this (they extend at same time, was hard to shot this screenshot xD):
2020-12-29_12h40_05
- [x] Dispensers/Droppers shouldn't dispense/drop on redstone off
- [x] Droppers should transfer items in containers
  - [x] Create new DispenseBehavior
  - [x] Implement in class
- [x] Doors shouldn't close, if unpowered redstone is placed beside them
- [x] If you place a door/fence gate/trapdoor in front of observer, lay out redstone on the output of the observer (**don't connect to door**) and then open the door, it immediately closes. (May be related to the issue above)
- [x] Create same as descibed above, but use dropper instead of door. When powering the dropper, it will spend out all his items like a clock and the observer output blinks. (was related to the other dropper bug and maybe to the door bug above)
- [x] The top piston on the image should retract, when redstone is powered off. (A blockupdate after the redstone was powered off makes it retract)
2020-11-10_17h33_19 2020-11-10_17h33_33
- [x] Locked Hoppers don't accept new items, but they also shouldn't give out Items
- [X] Observer's output should strongly power the block right at the output
- [x] Add dropper and dispenser sounds
- [x] Some redstone power sources have wrong usage of `Level#updateAroundRedstone()`
- [x] Lamps didn't receive power threw blocks
- [X] Pistons don't retract honey blocks in vertical direction
- [X] Piston push limit is wrong when using honey/slime blocks. 
- [X] Fix wrong repeater delay
- [x] It is weird, but vanilla:
2020-12-28_19h04_35
- [x] Another weird vanilla behaviour (why does the lamp burn?)
2020-12-28_20h08_09
- [x] When giving a one tick pulse to a piston (with blocks attached), the new position of the blocks don't receive a redstone update
- [x] Honey and slime block connected in piston facing direction on retract
- [X] Added an interface (`RedstoneComponent`) to all redstone components and moved redstone updates there
- [X] Created `#isGettingPower()` method in block for easy checking if a redstone component block (or any other) is powered

### CODE REVIEW

**Comment by PocketMiner82** [comment]
I just figured out, that you can link IntelliJ IDEA with github repos xD Before I always copy pasted the code from IntelliJ to github website editor :d

**Comment by PocketMiner82** [comment]
Does anyone of you know, how to check, if a door was opened and/or closed manually (by a player, not by redstone)? Saving in a local var doesn't work, as the class gets unloaded, if no longer needed...

**Comment by PocketMiner82** [comment]
From now, this design for a 3x3 piston door is working in PN 😁: 
https://www.youtube.com/watch?v=CnmDVRgS3-I

**Comment by PocketMiner82** [comment]
*Note: BlockBell needs update to BlockProperties.*

**Comment by PocketMiner82** [comment]
I'm currently doing some ingame test, then I'll mark as ready for review.

**Comment by PocketMiner82** [comment]
As said before, the Bell needs an update to BlockProperties to work with redstone, I don't know how to do this -.-

**Comment by PocketMiner82** [comment]
@LoboMetalurgico could you change to Status: Awaiting Review?

**Comment by LoboMetalurgico** [comment]
Of Course!

**Comment by PocketMiner82** [comment]
Review? xD

**Comment by PocketMiner82** [comment]
Looks like @joserobjr already updated BlockBell to blockproperties in #697; nice :D

**Comment by PocketMiner82** [comment]
@LoboMetalurgico could you link those two issues please?

**Comment by LoboMetalurgico** [comment]
Done!

**Comment by PocketMiner82** [comment]
@LoboMetalurgico do you know, when this will get reviewed? only wanna know whats going on currently, I don't want to hurry you or jose xD

**Comment by LoboMetalurgico** [comment]
Its pull proposes major changes and corrections, this makes the revision take a little longer than usual. But don't worry, I don't think it will take long.

If you want to have a better contact with us, you can join our discord: https://discord.gg/JVwf39q

**Comment by PocketMiner82** [comment]
Ok, sounds good :D

unfortunately I don't have a discord account atm, but thanks for the invite

**Comment by PocketMiner82** [comment]
Converted to draft because I want to continue finding some bugs on redstone (found a triangular piston door design on youtube, wanna make it work here).

**Comment by PocketMiner82** [comment]
This works now: https://www.youtube.com/watch?v=_Xw8skFnwi4
And this too: https://www.youtube.com/watch?v=YqpdhimKc8Q

**Comment by MrRazamataz** [comment]
Madlad. Good job!

**Comment by PocketMiner82** [comment]
Marking as ready for review, because I can't find more to fix xD

**Comment by MrRazamataz** [comment]
> Marking as ready for review, because I can't find more to fix xD
> Edit: I just found one, lol xD

what about minecarts despawning ?

**Comment by PocketMiner82** [comment]
> > Marking as ready for review, because I can't find more to fix xD
> > Edit: I just found one, lol xD
> 
> what about minecarts despawning ?

If I would do it, then in a separate PR which is about minecarts and rails, etc. but first, this PR has to be merged.

**Comment by RORIdev** [comment]
All the functionality on that initial message was confirmed. I'll now test for the next messages.

Note to author : This PR took  a *while* to even be processed because it is **_massive_**, the changes on the PR are sure *welcome* but try splitting the pull requests into smaller parts later. It'll make the reviewer's lives' easier and the pr will be accepted faster.

**Comment by RORIdev** [comment]
Everything seems fine. I'll review the code and merge if i can't find anything missing

File: src/main/java/cn/nukkit/block/BlockBell.java [hunk 1]
```diff
@@ -312,6 +314,24 @@ public int onUpdate(int type) {
         return 0;
     }
 
+    @Override
+    @PowerNukkitOnly
+    public boolean isGettingPower() {
```
Comment (RORIdev) [inline src/main/java/cn/nukkit/block/BlockBell.java:319]: `PowerNukkitSince` annotation missing

File: src/main/java/cn/nukkit/block/BlockDoor.java [hunk 2]
```diff
@@ -33,26 +35,29 @@
  */
 public abstract class BlockDoor extends BlockTransparentMeta implements Faceable {
     private static final double THICKNESS = 3.0 /16;
-    
+
+    // no idea how to make this better :d
+    private static final List<Location> manualOverrides = new ArrayList<>();
```
Comment (RORIdev) [inline src/main/java/cn/nukkit/block/BlockDoor.java:40]: After reading the whole diff for `BlockDoor`, I'm still lost on what this `manualOverrides` field do. Can you clarify it for me?

File: src/main/java/cn/nukkit/block/BlockDoor.java [hunk 3]
```diff
@@ -153,27 +158,101 @@ private void onNormalUpdate() {
...
+    }
+
+    @Override
+    @PowerNukkitOnly
+    @Since("1.4.0.0-PN")
+    public boolean isGettingPower() {
+        Location down;
+        Location up;
+        if (this.isTop()) {
+            down = down().getLocation().clone();
```
Comment (RORIdev) [inline src/main/java/cn/nukkit/block/BlockDoor.java:233]: Why you need to deep copy the location of the block? wouldn't just using the reference be enough for this use case?

**Review by RORIdev — CHANGES_REQUESTED** [review]

**Comment by MrRazamataz** [comment]
> > > Marking as ready for review, because I can't find more to fix xD
> > > Edit: I just found one, lol xD
> > 
> > 
> > what about minecarts despawning ?
> 
> If I would do it, then in a separate PR which is about minecarts and rails, etc. but first, this PR has to be merged.

Noone is gonna be able to use this becasue right now, the latest bleeding branch, doesnt let people join the server. It may in a clean setup but most if not all people are having trouble with even getting people to join on the latest bleeding branch.

**Comment by PocketMiner82** [comment]
I can join, lol xD

**Comment by PocketMiner82** [comment]
I'll look in to the requests to change tomorrow and I'll split such PRs in future, thx for information xD @RORIdev

**Comment by MrRazamataz** [comment]
> I can join, lol xD

Alot of people are having issues with it. The latest "safe" one is: https://github.com/PowerNukkit/PowerNukkit/actions/runs/434067711


I think

File: src/main/java/cn/nukkit/block/BlockDoor.java (hunk 2 above)
Comment (PocketMiner82) [reply src/main/java/cn/nukkit/block/BlockDoor.java:40]: It contains a list of positions of Doors, which have been opened by hand (by a player). It is used to detect on redstone update, if the door should be closed if redstone is off on the update, previously the door always closed, when placing an unpowered redstone at the door, this fixes it and gives the vanilla behavior.

File: src/main/java/cn/nukkit/block/BlockDoor.java (hunk 3 above)
Comment (PocketMiner82) [reply src/main/java/cn/nukkit/block/BlockDoor.java:233]: yes would be, I'll edit it

File: src/main/java/cn/nukkit/block/BlockDoor.java (hunk 2 above)
Comment (PocketMiner82) [reply src/main/java/cn/nukkit/block/BlockDoor.java:43]: I added a comment describing it.

File: src/main/java/cn/nukkit/block/BlockBell.java (hunk 1 above)
Comment (PocketMiner82) [reply src/main/java/cn/nukkit/block/BlockBell.java:319]: Added.

**Comment by PocketMiner82** [comment]
> > I can join, lol xD
> 
> Alot of people are having issues with it. The latest "safe" one is: https://github.com/PowerNukkit/PowerNukkit/actions/runs/434067711
> 
> I think

@MrRazamataz Have you created an issue? Does it show any error? Are you using WaterDog?

**Comment by MrRazamataz** [comment]
All it said was they disconnected or timed out. I'm not using waterdog or any proxy and there was no error. I will try to recreate the issue.

**Comment by PocketMiner82** [comment]
Hmm... Weird.

**Comment by PocketMiner82** [comment]
Maybe you could try a version from this PR, idk xD
https://github.com/PowerNukkit/PowerNukkit/suites/1755062552/artifacts/33393538

**Comment by RealTriassic** [comment]
The issue with disconnecting on latest bleeding also happens with Waterdog.

**Comment by MonoAdrian23** [comment]
It happens without Waterdog and Waterdog

**Review by RORIdev — CHANGES_REQUESTED** [review]
Some other changes are written here.  
I also noted a lot of code duplication on the redstone ` updateAroundX` logic. Would it be a good idea to make an `updateAround` function on a base class common to all redstone components, and that be called instead? since afaik when you need to send a redstone update  from a to b, a just calls a bunch of blockUpdates to the blocks around and it functions on a similar way. just a minor code quality thing.

**Comment by PocketMiner82** [comment]
> Some other changes are written here.  
> I also noted a lot of code duplication on the redstone ` updateAroundX` logic. Would it be a good idea to make an `updateAround` function on a base class common to all redstone components, and that be called instead? since afaik when you need to send a redstone update  from a to b, a just calls a bunch of blockUpdates to the blocks around and it functions on a similar way. just a minor code quality thing.

@RORIdev You mean like the `#isGettingPowered()` in `Block`?

**Comment by RORIdev** [comment]
> 
> 
> @RORIdev You mean like the `#isGettingPowered()` in `Block`?

something along these lines, yeah. but on a base class common to redstone components, if it exists. if not just make a BlockRedstone or somthing like that.

**Comment by PocketMiner82** [comment]
for my understanding: something like a utility class for redstone related stuff? if so, should I also move the isGettingPowered to that?

**Comment by RORIdev** [comment]
> for my understanding: something like a utility class for redstone related stuff? if so, should I also move the isGettingPowered to that?
Yes. thats is, but normal blocks *can* be powered by redstone so `isGettingPowered` shouldn't be moved.

**Comment by PocketMiner82** [comment]
Ok, I'll do that. A last question: Should the class be extended by redstone components (like `SomeClass extends RedstoneComponent`) or should it contain static methods and be in the `cn.nukkit.utils` package

**Comment by PocketMiner82** [comment]
Both has positive aspects: With the parent class, plugins could easy check if a block is a redstone component, but with static methods, plugins could also use to redstone update non redstone components.

File: src/main/java/cn/nukkit/block/BlockDoor.java [hunk 4]
```diff
@@ -153,42 +161,111 @@ private void onNormalUpdate() {
...
+        }
+    }
+
+    @PowerNukkitOnly
+    @Since("1.4.0.0-PN")
+    public boolean getManualOverride() {
+        Location down;
+        Location up;
+        if (this.isTop()) {
+            down = down().getLocation().clone();
```
[the remark this answers is not in the document]
Comment (PocketMiner82) [reply src/main/java/cn/nukkit/block/BlockDoor.java]: Done

File: src/main/java/cn/nukkit/block/BlockDoor.java [hunk 5]
```diff
@@ -153,42 +161,111 @@ private void onNormalUpdate() {
...
         }
     }
 
+    @PowerNukkitOnly
+    @Since("1.4.0.0-PN")
+    public void setManualOverride(boolean val) {
+        Location down;
+        Location up;
+        if (this.isTop()) {
+            down = down().getLocation().clone();
```
[the remark this answers is not in the document]
Comment (PocketMiner82) [reply src/main/java/cn/nukkit/block/BlockDoor.java]: Done

**Comment by RORIdev** [comment]
> Both have positive aspects: With the parent class, plugins could easy check if a block is a redstone component, but with static methods, plugins could also use to redstone update non redstone components.

Why not both? I can't see why those options are exclusive at all.

**Comment by PocketMiner82** [comment]
I'll try 😀

**Comment by PocketMiner82** [comment]
Oh; most classes already extend BlockSolidMeta or BlockTransparentMeta...

**Comment by RORIdev** [comment]
> Oh; most classes already extend BlockSolidMeta or BlockTransparentMeta or BlockFlowable...

Can't you make something along these lines in java?
```cs
abstract class RedstoneComponent : Block {}
class RedstoneDust : RedstoneComponent, BlockTransparent {}
```
then `RedstoneDust` will inherit all common properties and functions from `RedstoneComponent`, while still being a BlockTransparent.

iirc thats how `BlockTrapdoor` and its derivatives [`BlockAcaciaTrapdoor`, `BlockSpruceTrapdoor`, ...] work.

**Comment by PocketMiner82** [comment]
We can't extend multiple classes, but I could try with an interface...

**Comment by RORIdev** [comment]
> We can't extend multiple classes, but I could try with an interface...

That should do the trick

**Comment by PocketMiner82** [comment]
```java
package cn.nukkit.block;

/**
 * Interface containing redstone related methods.
 */
public interface RedstoneComponent {
    static void updateAroundRedstone() {

    }
}
```

And:
```java
public abstract class BlockRedstoneComparator extends BlockRedstoneDiode implements RedstoneComponent, BlockEntityHolder<BlockEntityComparator> {
```

Just to make sure; this is what you mean?

And I can't call the static method directly, I have to use RedstoneComponent before it.

**Comment by RORIdev** [comment]
Yes. that's perfect.

**Comment by PocketMiner82** [comment]
Should I also remove the redstone update methods from level and move it here?

**Comment by RORIdev** [comment]
> Should I also ~remove~ depract the redstone update methods from level and move it here?

If you don't modify existing public api methods, sure. They can go live there.

**Comment by PocketMiner82** [comment]
@RORIdev do you think it's better when also redstone consumers implement the RedstoneComponent interface or just the power sources (and then maybe rename the interface to RedstonePowerSource) or should there be three interfaces:
RedstoneComponent (parent class, implementing all the update around stuff), RedstonePowerSource (implementing RedstoneComponent) and RedstoneConsumer (also implementing RedstoneComponent)?

**Comment by RORIdev** [comment]
> @RORIdev do you think it's better when also redstone consumers implement the RedstoneComponent interface or just the power sources (and then maybe rename the interface to RedstonePowerSource) or should there be three interfaces:
> RedstoneComponent (parent class, implementing all the update around stuff), RedstonePowerSource (implementing RedstoneComponent) and RedstoneConsumer (also implementing RedstoneComponent)?
> 
> Edit: ~I'd prefer the last one :P~ Nope, not possible (e.g. repeater is consumer and source...), so the first one is the one I'd prefer xD

since the second one isn't possible, the first one is already good

**Comment by PocketMiner82** [comment]
Ok, so all redstone components (also consumers) will implement it.

**Comment by PocketMiner82** [comment]
Everything seems to be working as before ingame :)

**Review by RORIdev — APPROVED** [review]

**Review by RORIdev — COMMENTED** [review]
Thanks for your PR. I've reviewed the code and it looks great. I'll have to pass it to another reviewer for double checking purposes since this PR is massive.

As far as I could check nothing is weird, should be merged soon after another review is made.

**Comment by PocketMiner82** [comment]
I just found another bug: Glazed terracotta should be able to get pulled, only pushed by pistons; can I fix it here?

**Comment by RORIdev** [comment]
> I just found another bug: Glazed terracotta should not be able to get pulled, only pushed by pistons; can I fix it here?

This PR is already huge, so if you can make another one that'll be great. If this fix depends on this pr being merged, make the base of the new branch the last commit of this pr, and add a `Depends on : #750` on the description of the new PR.

**Comment by PocketMiner82** [comment]
Ok, I'll do :)

**Comment by PocketMiner82** [comment]
[world.zip](https://github.com/PowerNukkit/PowerNukkit/files/5766005/world.zip)
This is my powernukkit testworld with the piston doors, if someone needs xD

**Comment by PocketMiner82** [comment]
I'll delete my account soon, as I want a new email and username.

If you want to keep my work in a PR I think you have to clone the branch to PN.

Thanks,
PocketMiner82

**Comment by PocketMiner82** [comment]
Oh lol, thx for merge <3

**Comment by LoboMetalurgico** [comment]
Thank you for the incredible PR <3

**Comment by hteppl** [comment]
> I'll delete my account soon, as I want a new email and username.
> 
> If you want to keep my work in a PR I think you have to clone the branch to PN.
> 
> Thanks,
> PocketMiner82

you can do it without account changin lul

**Comment by PocketMiner82** [comment]
> > I'll delete my account soon, as I want a new email and username.
> > 
> > If you want to keep my work in a PR I think you have to clone the branch to PN.
> > 
> > Thanks,
> > PocketMiner82
> 
> you can do it without account changin lul

thx, I didn't know you can also change your email, I'll do it like you suggested.

**Pull request merged**